Ah, there is also a MSVC 64bits build failure in the testsuite:

```
runtime/runtime_events.c(289): error C2143: syntax error: missing ';' before '}'
runtime/runtime_events.c(282): warning C4133: 'initializing': incompatible types - from 'wchar_t *' to 'char *'
```

(The wchar_t thing might suggest that `char_os` should be used somewhere instead of `char`, I don't know.)

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -194,15 +194,17 @@ jobs:
         run: |
           apt-get update
           apt-get install -y git gcc make parallel
+          adduser --disabled-password --gecos '' ocaml
       - name: Checkout
         # See https://github.com/actions/checkout/issues/334
         uses: actions/checkout@v1
       - name: configure tree
         run: |
-          MAKE_ARG=-j bash -xe tools/ci/actions/runner.sh configure
+          chown -R ocaml:ocaml .
+          MAKE_ARG=-j su ocaml -c "bash -xe tools/ci/actions/runner.sh configure"
       - name: Build
         run: |
-          MAKE_ARG=-j bash -xe tools/ci/actions/runner.sh build
+          MAKE_ARG=-j su ocaml -c "bash -xe tools/ci/actions/runner.sh build"
       - name: Run the testsuite
         run: |
-          bash -xe tools/ci/actions/runner.sh test
+          su ocaml -c "bash -xe tools/ci/actions/runner.sh test"

--- a/Changes
+++ b/Changes
@@ -61,6 +61,10 @@ ___________
 
 ### Runtime system:
 
+- #13419: Fix memory bugs in runtime events system.
+  (B. Szilvasy and Nick Barnes, review by Miod Vallat, Nick Barnes,
+   Tim McGilchrist, and Gabriel Scherer)
+
 - #13364: Emit major slice counters in the runtime events.
   (KC Sivaramakrishnan and Sadiq Jaffer, review by Gabriel Scherer)
 

--- a/otherlibs/runtime_events/runtime_events.ml
+++ b/otherlibs/runtime_events/runtime_events.ml
@@ -401,6 +401,7 @@ end
 external start : unit -> unit = "caml_ml_runtime_events_start"
 external pause : unit -> unit = "caml_ml_runtime_events_pause"
 external resume : unit -> unit = "caml_ml_runtime_events_resume"
+external path : unit -> string option = "caml_ml_runtime_events_path"
 
 external create_cursor : (string * int) option -> cursor
                                         = "caml_ml_runtime_events_create_cursor"

--- a/otherlibs/runtime_events/runtime_events.mli
+++ b/otherlibs/runtime_events/runtime_events.mli
@@ -617,6 +617,10 @@ val start : unit -> unit
   a set of callbacks to be called for each type of event.
 *)
 
+val path : unit -> string option
+(** If runtime events are being collected, [path ()] returns [Some p] where [p]
+  is a path to the runtime events file. Otherwise, it returns None. *)
+
 val pause : unit -> unit
 (** [pause ()] will pause the collection of events in the runtime.
    Traces are collected if the program has called [Runtime_events.start ()] or

--- a/otherlibs/runtime_events/runtime_events_consumer.c
+++ b/otherlibs/runtime_events/runtime_events_consumer.c
@@ -89,104 +89,102 @@ struct caml_runtime_events_cursor {
 
 /* C-API for reading from an runtime_events */
 
-runtime_events_error
-caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
-                             struct caml_runtime_events_cursor **cursor_res) {
+/** Creates a new string with the path of the ring file. Returns a
+ * value from the runtime_events_error enum */
+static int format_runtime_ring_file(
+  const char_os *input_path, int input_pid,
+  char_os **out_ring_file
+) {
+  char_os *ring_file;
   int ret;
 
-#ifndef _WIN32
-  int ring_fd;
-  struct stat tmp_stat;
-#endif
-
-  struct caml_runtime_events_cursor *cursor =
-      caml_stat_alloc_noexc(sizeof(struct caml_runtime_events_cursor));
-  char_os *runtime_events_loc;
-
-  if (cursor == NULL) {
-    return E_ALLOC_FAIL;
-  }
-
-  runtime_events_loc = caml_stat_alloc_noexc(RING_FILE_NAME_MAX_LEN);
-
-  if (runtime_events_loc == NULL) {
-    caml_stat_free(cursor);
-    return E_ALLOC_FAIL;
-  }
-
-  /* If pid < 0 then we create a cursor for the current process */
-  if (pid < 0) {
-    runtime_events_loc = caml_runtime_events_current_location();
-
-    if( runtime_events_loc == NULL ) {
-      caml_stat_free(cursor);
-      return E_NO_CURRENT_RING;
+  if (input_pid < 0) {
+    /* Attaching to this process' ring, if it exists */
+    ring_file = caml_runtime_events_current_location();
+    if (ring_file == NULL) {
+      ret = E_NO_CURRENT_RING; /* could also be allocation failure */
+      goto fail_current_file;
     }
   } else {
-  /* In this case we are reading the ring for a different process */
-    if (runtime_events_path) {
-      ret = snprintf_os(runtime_events_loc, RING_FILE_NAME_MAX_LEN,
-                        T("%s/%d.events"), runtime_events_path, pid);
+    /* Attaching to a process by directory and PID */
+    int err;
+    ring_file = caml_stat_alloc_noexc(RING_FILE_NAME_MAX_LEN);
+    if (ring_file == NULL) {
+      ret = E_ALLOC_FAIL;
+      goto fail_alloc_file;
+    }
+
+    if (input_path) {
+      err = snprintf_os(ring_file, RING_FILE_NAME_MAX_LEN,
+                        T("%s/%d.events"), input_path, input_pid);
     } else {
-      ret =
-          snprintf_os(runtime_events_loc, RING_FILE_NAME_MAX_LEN,
-                      T("%d.events"), pid);
+      err = snprintf_os(ring_file, RING_FILE_NAME_MAX_LEN,
+                        T("%d.events"), input_pid);
     }
-
-    if (ret < 0) {
-      caml_stat_free(cursor);
-      caml_stat_free(runtime_events_loc);
-      return E_PATH_FAILURE;
+    if (err < 0) {
+      ret = E_PATH_FAILURE;
+      goto fail_snprintf;
     }
   }
 
+  *out_ring_file = ring_file;
+  return E_SUCCESS;
+
+ fail_snprintf:
+  caml_stat_free(ring_file);
+ fail_alloc_file:
+ fail_current_file:
+  return ret;
+}
+
+/* Creates and maps the ring file. Returns a value from the
+ * runtime_events_error enum */
+
+static int
+cursor_map_ring_file(struct caml_runtime_events_cursor *cursor,
+                     char_os *ring_file)
+{
+  int ret = 0;
 #ifdef _WIN32
-  cursor->ring_file_handle = CreateFile(
-    runtime_events_loc,
-    GENERIC_READ | GENERIC_WRITE,
-    FILE_SHARE_READ | FILE_SHARE_WRITE,
-    NULL,
-    OPEN_EXISTING,
-    FILE_ATTRIBUTE_NORMAL,
-    NULL
-  );
-
-  if (cursor->ring_file_handle == INVALID_HANDLE_VALUE) {
-    caml_stat_free(cursor);
-    caml_stat_free(runtime_events_loc);
-    return E_OPEN_FAILURE;
+  HANDLE ring_file_handle = CreateFile(ring_file,
+                                       GENERIC_READ | GENERIC_WRITE,
+                                       FILE_SHARE_READ | FILE_SHARE_WRITE,
+                                       NULL,
+                                       OPEN_EXISTING,
+                                       FILE_ATTRIBUTE_NORMAL,
+                                       NULL);
+  if (ring_file_handle == INVALID_HANDLE_VALUE) {
+    ret = E_OPEN_FAILURE;
+    goto fail_create_file;
   }
 
-  cursor->ring_handle = CreateFileMapping(
-    cursor->ring_file_handle,
-    NULL,
-    PAGE_READWRITE,
-    0,
-    0,
-    NULL
-  );
-
-  if (cursor->ring_handle == INVALID_HANDLE_VALUE) {
-    caml_stat_free(cursor);
-    caml_stat_free(runtime_events_loc);
-    return E_MAP_FAILURE;
+  HANDLE ring_handle = CreateFileMapping(ring_file_handle, NULL,
+                                         PAGE_READWRITE, 0, 0, NULL);
+  if (ring_handle == INVALID_HANDLE_VALUE) {
+    ret = E_MAP_FAILURE;
+    goto fail_create_mapping;
   }
 
-  cursor->map = MapViewOfFile(
-    cursor->ring_handle,
-    FILE_MAP_ALL_ACCESS,
-    0,
-    0,
-    0
-  );
+  void *map = MapViewOfFile(ring_handle, FILE_MAP_ALL_ACCESS,
+                            0, 0, 0);
 
-  if( cursor->map == NULL ) {
-    caml_stat_free(cursor);
-    caml_stat_free(runtime_events_loc);
-    return E_MAP_FAILURE;
+  if( map == NULL ) {
+    ret = E_MAP_FAILURE;
+    goto fail_map_view;
   }
 
+  cursor->ring_file_handle = ring_file_handle;
+  cursor->ring_handle = ring_handle;
+  cursor->map = map;
   cursor->ring_file_size_bytes = GetFileSize(cursor->ring_file_handle, NULL);
+  return E_SUCCESS;
+
+ fail_map_view:
+  CloseHandle(ring_handle);
+ fail_create_mapping:
+  CloseHandle(ring_file_handle);
+ fail_create_file:
+  return ret;
 #else
 #if defined(__ARM_ARCH) && __ARM_ARCH <= 5
   /* Atomic 64-bit load requires RW memory on Debian armel.  See:
@@ -197,69 +195,117 @@ caml_runtime_events_create_cursor(const char_os* runtime_events_path, int pid,
   const int open_flags = O_RDONLY;
   const int mmap_prot = PROT_READ;
 #endif
-  ring_fd = open(runtime_events_loc, open_flags, 0);
-
-  if( ring_fd == -1 ) {
-    caml_stat_free(cursor);
-    caml_stat_free(runtime_events_loc);
-    return E_OPEN_FAILURE;
+  int ring_fd = open(ring_file, open_flags, 0);
+  if(ring_fd == -1) {
+    ret = E_OPEN_FAILURE;
+    goto fail_open;
   }
 
+  struct stat tmp_stat;
   ret = fstat(ring_fd, &tmp_stat);
-
   if (ret < 0) {
-    caml_stat_free(cursor);
-    caml_stat_free(runtime_events_loc);
-    return E_OPEN_FAILURE;
+    ret = E_OPEN_FAILURE;
+    goto fail_fstat;
   }
-
-  cursor->ring_file_size_bytes = tmp_stat.st_size;
+  size_t ring_file_size_bytes = tmp_stat.st_size;
 
   /* This cast is necessary for compatibility with Illumos' non-POSIX
     mmap/munmap */
-  cursor->map = (void*) mmap(NULL, cursor->ring_file_size_bytes, mmap_prot,
-                          MAP_SHARED, ring_fd, 0);
+  void *map = (void*) mmap(NULL, ring_file_size_bytes, mmap_prot,
+                           MAP_SHARED, ring_fd, 0);
 
-  if( cursor->map == MAP_FAILED ) {
-    caml_stat_free(cursor);
-    caml_stat_free(runtime_events_loc);
-    return E_MAP_FAILURE;
+  if( map == MAP_FAILED ) {
+    ret = E_MAP_FAILURE;
+    goto fail_map;
   }
+
+  (void)close(ring_fd);
+  cursor->map = map;
+  cursor->ring_file_size_bytes = ring_file_size_bytes;
+  return E_SUCCESS;
+
+ fail_map:
+ fail_fstat:
+  (void)close(ring_fd);
+ fail_open:
+  return ret;
 #endif
+}
+
+/* unmaps the ring file from a cursor */
+static void cursor_unmap_ring_file(struct caml_runtime_events_cursor *cursor)
+{
+#ifdef _WIN32
+  UnmapViewOfFile(cursor->map);
+  CloseHandle(cursor->ring_file_handle);
+  CloseHandle(cursor->ring_handle);
+#else
+  munmap(cursor->map, cursor->ring_file_size_bytes);
+#endif
+}
+
+runtime_events_error caml_runtime_events_create_cursor(
+  const char_os* runtime_events_path, int pid,
+  struct caml_runtime_events_cursor **cursor_res
+) {
+  int ret = E_SUCCESS;
+
+  struct caml_runtime_events_cursor *cursor =
+    caml_stat_alloc_noexc(sizeof(struct caml_runtime_events_cursor));
+  if (cursor == NULL) {
+    ret = E_ALLOC_FAIL;
+    goto fail_alloc_cursor;
+  }
+  /* zero out all fields, notably the callbacks */
+  memset(cursor, 0, sizeof(*cursor));
+
+  char_os *ring_file;
+  ret = format_runtime_ring_file(runtime_events_path, pid,
+                                 &ring_file);
+  if (ret != E_SUCCESS) {
+    goto fail_format_file;
+  }
+
+  ret = cursor_map_ring_file(cursor, ring_file);
+  if (ret != E_SUCCESS) {
+    goto fail_map_ring_file;
+  }
 
   cursor->metadata = *(struct runtime_events_metadata_header*)cursor->map;
 
   if (cursor->metadata.max_domains > Max_domains_max) {
-    /* avoid overflow in multiplication below */
-    caml_stat_free(cursor);
-    caml_stat_free(runtime_events_loc);
-    return E_CORRUPT_STREAM;
+    ret = E_CORRUPT_STREAM;
+    goto fail_metadata_corrupt;
   }
 
   cursor->current_positions =
-      caml_stat_alloc(cursor->metadata.max_domains * sizeof(uint64_t));
+      caml_stat_alloc_noexc(cursor->metadata.max_domains * sizeof(uint64_t));
+  if (cursor->current_positions == NULL) {
+    ret = E_ALLOC_FAIL;
+    goto fail_current_pos;
+  }
 
   for (int j = 0; j < cursor->metadata.max_domains; j++) {
     cursor->current_positions[j] = 0;
   }
+
   cursor->cursor_open = 1;
   atomic_store(&cursor->cursor_in_poll, 0);
   cursor->next_read_domain = 0;
-
-  cursor->runtime_begin = NULL;
-  cursor->runtime_end = NULL;
-  cursor->runtime_counter = NULL;
-  cursor->alloc = NULL;
-  cursor->lifecycle = NULL;
-  cursor->lost_events = NULL;
-  cursor->user_unit = NULL;
-  cursor->user_int = NULL;
-  cursor->user_span = NULL;
-  cursor->user_custom = NULL;
-
   *cursor_res = cursor;
+  caml_stat_free(ring_file);
 
   return E_SUCCESS;
+
+ fail_current_pos:
+ fail_metadata_corrupt:
+  cursor_unmap_ring_file(cursor);
+ fail_map_ring_file:
+  caml_stat_free(ring_file);
+ fail_format_file:
+  caml_stat_free(cursor);
+ fail_alloc_cursor:
+  return ret;
 }
 
 void caml_runtime_events_set_runtime_begin(
@@ -353,15 +399,7 @@ void caml_runtime_events_set_user_custom(
 void caml_runtime_events_free_cursor(struct caml_runtime_events_cursor *cursor){
   if (cursor->cursor_open) {
     cursor->cursor_open = 0;
-#ifdef _WIN32
-    UnmapViewOfFile(cursor->map);
-    CloseHandle(cursor->ring_file_handle);
-    CloseHandle(cursor->ring_handle);
-#else
-    /* This cast is necessary for compatibility with Illumos' non-POSIX
-      mmap/munmap */
-    munmap((void*)cursor->map, cursor->ring_file_size_bytes);
-#endif
+    cursor_unmap_ring_file(cursor);
     caml_stat_free(cursor->current_positions);
     caml_stat_free(cursor);
   }
@@ -1192,11 +1230,11 @@ CAMLprim value caml_ml_runtime_events_create_cursor(value path_pid_option) {
 
   res = caml_runtime_events_create_cursor(path, pid, &cursor);
 
-  if (res != E_SUCCESS) {
-    if( path != NULL ) {
-      caml_stat_free(path);
-    }
+  if( path != NULL ) {
+    caml_stat_free(path);
+  }
 
+  if (res != E_SUCCESS) {
     switch(res) {
       case E_PATH_FAILURE:
         caml_failwith(
@@ -1228,10 +1266,6 @@ CAMLprim value caml_ml_runtime_events_create_cursor(value path_pid_option) {
   caml_runtime_events_set_user_custom(cursor, ml_user_custom);
 
   Cursor_val(wrapper) = cursor;
-
-  if( path != NULL ) {
-    caml_stat_free(path);
-  }
 
   // 3 words block:
   //  - cursor
@@ -1303,4 +1337,4 @@ CAMLprim value caml_ml_runtime_events_read_poll(value wrapper,
   }
 
   CAMLreturn(Val_int(events_consumed));
-};
+}

--- a/runtime/caml/memory.h
+++ b/runtime/caml/memory.h
@@ -140,14 +140,20 @@ typedef char* caml_stat_string;
    the request fails, and so requires the runtime lock to be held.
 */
 CAMLextern caml_stat_string caml_stat_strdup(const char *s);
-#ifdef _WIN32
-CAMLextern wchar_t* caml_stat_wcsdup(const wchar_t *s);
-#endif
 
 /* [caml_stat_strdup_noexc] is a variant of [caml_stat_strdup] that returns NULL
    in case the request fails, and doesn't require the runtime lock.
 */
 CAMLextern caml_stat_string caml_stat_strdup_noexc(const char *s);
+
+#ifdef _WIN32
+/* On Windows, [caml_stat_wcsdup] and [caml_stat_wcsdup_noexc] are the
+ * obvious equivalents of [caml_stat_strdup] and
+ * [caml_stat_strdup_noexc] for wide characters.
+ */
+CAMLextern wchar_t* caml_stat_wcsdup(const wchar_t *s);
+CAMLextern wchar_t* caml_stat_wcsdup_noexc(const wchar_t *s);
+#endif
 
 /* [caml_stat_strconcat(nargs, strings)] concatenates null-terminated [strings]
    (an array of [char*] of size [nargs]) into a new string, dropping all NULs,

--- a/runtime/caml/misc.h
+++ b/runtime/caml/misc.h
@@ -455,10 +455,12 @@ extern double caml_log1p(double);
 #define clock_os caml_win32_clock
 
 #define caml_stat_strdup_os caml_stat_wcsdup
+#define caml_stat_strdup_noexc_os caml_stat_wcsdup_noexc
 #define caml_stat_strconcat_os caml_stat_wcsconcat
 
 #define caml_stat_strdup_to_os caml_stat_strdup_to_utf16
 #define caml_stat_strdup_of_os caml_stat_strdup_of_utf16
+#define caml_stat_strdup_noexc_of_os caml_stat_strdup_noexc_of_utf16
 #define caml_copy_string_of_os caml_copy_string_of_utf16
 
 #else /* _WIN32 */
@@ -495,10 +497,12 @@ extern double caml_log1p(double);
 #define clock_os clock
 
 #define caml_stat_strdup_os caml_stat_strdup
+#define caml_stat_strdup_noexc_os caml_stat_strdup_noexc
 #define caml_stat_strconcat_os caml_stat_strconcat
 
 #define caml_stat_strdup_to_os caml_stat_strdup
 #define caml_stat_strdup_of_os caml_stat_strdup
+#define caml_stat_strdup_noexc_of_os caml_stat_strdup_noexc
 #define caml_copy_string_of_os caml_copy_string
 
 #endif /* _WIN32 */

--- a/runtime/caml/osdeps.h
+++ b/runtime/caml/osdeps.h
@@ -169,12 +169,25 @@ extern void caml_init_os_params(void);
 */
 CAMLextern wchar_t* caml_stat_strdup_to_utf16(const char *s);
 
+/* [caml_stat_strdup_noexc_of_utf16(s)] returns a null-terminated copy of [s],
+   re-encoded in UTF-8 if [caml_windows_unicode_runtime_enabled] is non-zero or
+   the current Windows code page otherwise.
+
+   The returned string is allocated with [caml_stat_alloc_noexc], so
+   it should be freed using [caml_stat_free].
+
+   If allocation fails, this returns NULL.
+*/
+CAMLextern char* caml_stat_strdup_noexc_of_utf16(const wchar_t *s);
+
 /* [caml_stat_strdup_of_utf16(s)] returns a null-terminated copy of [s],
    re-encoded in UTF-8 if [caml_windows_unicode_runtime_enabled] is non-zero or
    the current Windows code page otherwise.
 
-   The returned string is allocated with [caml_stat_alloc], so it should be free
-   using [caml_stat_free].
+   The returned string is allocated with [caml_stat_alloc_noexc], so
+   it should be freed using [caml_stat_free].
+
+   If allocation fails, this raises Out_of_memory.
 */
 CAMLextern char* caml_stat_strdup_of_utf16(const wchar_t *s);
 

--- a/runtime/caml/runtime_events.h
+++ b/runtime/caml/runtime_events.h
@@ -295,8 +295,10 @@ void caml_runtime_events_destroy(void);
    in a forked child */
 CAMLextern void caml_runtime_events_post_fork(void);
 
-/* Returns the location of the runtime_events for the current process if started
-   or NULL otherwise */
+/* Return the path of the ring buffers file of this process, or NULL
+   if runtime events are not enabled. This is used in the consumer to
+   read the ring buffers of the current process. Always returns a
+   freshly-allocated string. */
 CAMLextern char_os* caml_runtime_events_current_location(void);
 
 /* Functions for putting runtime data on to the runtime_events. These are all

--- a/runtime/memory.c
+++ b/runtime/memory.c
@@ -731,13 +731,21 @@ CAMLexport caml_stat_string caml_stat_strdup(const char *s)
 
 #ifdef _WIN32
 
-CAMLexport wchar_t * caml_stat_wcsdup(const wchar_t *s)
+CAMLexport wchar_t * caml_stat_wcsdup_noexc(const wchar_t *s)
 {
   size_t slen = wcslen(s);
   wchar_t* result = caml_stat_alloc((slen + 1)*sizeof(wchar_t));
   if (result == NULL)
-    caml_raise_out_of_memory();
+    return NULL;
   memcpy(result, s, (slen + 1)*sizeof(wchar_t));
+  return result;
+}
+
+CAMLexport wchar_t * caml_stat_wcsdup(const wchar_t *s)
+{
+  wchar_t* result = caml_stat_wcsdup_noexc(s);
+  if (result == NULL)
+    caml_raise_out_of_memory();
   return result;
 }
 

--- a/runtime/runtime_events.c
+++ b/runtime/runtime_events.c
@@ -205,11 +205,13 @@ void caml_runtime_events_post_fork(void) {
   }
 }
 
-/* Return the current location for the ring buffers of this process. This is
-  used in the consumer to read the ring buffers of the current process */
+/* Return the path of the ring buffers file of this process, or NULL
+   if runtime events are not enabled. This is used in the consumer to
+   read the ring buffers of the current process. Always returns a
+   freshly-allocated string. */
 char_os* caml_runtime_events_current_location(void) {
   if( atomic_load_acquire(&runtime_events_enabled) ) {
-    return current_ring_loc;
+    return caml_stat_strdup_noexc_os(current_ring_loc);
   } else {
     return NULL;
   }
@@ -278,9 +280,13 @@ static void runtime_events_create_from_stw_single(void) {
 
     if (ring_file_handle == INVALID_HANDLE_VALUE) {
       char* ring_loc_u8 = caml_stat_strdup_of_os(current_ring_loc);
-      caml_fatal_error("Couldn't open ring buffer loc: %s",
-                        ring_loc_u8);
-      caml_stat_free(ring_loc_u8);
+      if (ring_loc_u8) {
+        caml_fatal_error("Couldn't open ring buffer file: %s",
+                         ring_loc_u8);
+        caml_stat_free(ring_loc_u8);
+      } else {
+        caml_fatal_error("Couldn't open ring buffer file");
+      }
     }
 
     ring_handle = CreateFileMapping(
@@ -478,6 +484,23 @@ CAMLprim value caml_ml_runtime_events_pause(value vunit) {
 
 CAMLprim value caml_ml_runtime_events_resume(value vunit) {
   caml_runtime_events_resume(); return Val_unit;
+}
+
+CAMLprim value caml_ml_runtime_events_path(value vunit) {
+  CAMLparam0();
+  CAMLlocal1 (res);
+  res = Val_none;
+  if (atomic_load_acquire(&runtime_events_enabled)) {
+    res = caml_alloc_small(1, Tag_some);
+    /* The allocation might GC, which could allow another domain to
+     * nuke current_ring_loc, so we check again. */
+    if (atomic_load_acquire(&runtime_events_enabled)) {
+      Field(res, 0) = caml_copy_string_of_os(current_ring_loc);
+    } else {
+      res = Val_none;
+    }
+  }
+  CAMLreturn(res);
 }
 
 CAMLprim value caml_ml_runtime_events_are_active(void) {

--- a/runtime/win32.c
+++ b/runtime/win32.c
@@ -964,15 +964,25 @@ CAMLexport wchar_t* caml_stat_strdup_to_utf16(const char *s)
   return ws;
 }
 
-CAMLexport caml_stat_string caml_stat_strdup_of_utf16(const wchar_t *s)
+CAMLexport caml_stat_string caml_stat_strdup_noexc_of_utf16(const wchar_t *s)
 {
   caml_stat_string out;
   int retcode;
 
   retcode = caml_win32_wide_char_to_multi_byte(s, -1, NULL, 0);
-  out = caml_stat_alloc(retcode);
-  caml_win32_wide_char_to_multi_byte(s, -1, out, retcode);
+  out = caml_stat_alloc_noexc(retcode);
+  if (out != NULL) {
+    caml_win32_wide_char_to_multi_byte(s, -1, out, retcode);
+  }
 
+  return out;
+}
+
+CAMLexport caml_stat_string caml_stat_strdup_of_utf16(const wchar_t *s)
+{
+  caml_stat_string out = caml_stat_strdup_noexc_of_utf16(s);
+  if (out == NULL)
+    caml_raise_out_of_memory();
   return out;
 }
 

--- a/testsuite/tests/lib-runtime-events/test_create_cursor_failures.ml
+++ b/testsuite/tests/lib-runtime-events/test_create_cursor_failures.ml
@@ -1,0 +1,47 @@
+(* TEST
+ include unix;
+ include runtime_events;
+ hasunix;
+ {
+   bytecode;
+ }{
+   native;
+ }
+*)
+
+(* Tests that [create_cursor]:
+ * - fails on [None] if runtime events haven't been started
+ * - doesn't double-free when it fails to attach to [None]
+ * - does manage to attach to this process if we provide the right PID
+ *)
+
+let create_and_free ?(pid) () =
+  try
+    let dir_and_pid = Option.map (fun p -> ".", p) pid in
+    let cur = Runtime_events.create_cursor dir_and_pid in
+    Runtime_events.free_cursor cur;
+    print_endline "OK"
+  with Failure msg -> print_endline msg
+
+let start_and_pause () =
+  Runtime_events.start ();
+  Runtime_events.pause ()
+
+(* Windows workaround to get the correct PID *)
+let find_events_pid cursor =
+  Scanf.sscanf (Option.get (Runtime_events.path())) "%d.events" Fun.id
+
+(* force failure of [create_cursor None] *)
+let make_unreadable () =
+  Unix.chmod (Option.get (Runtime_events.path())) 0o000
+
+let () =
+  create_and_free (); (* fail, not started *)
+  start_and_pause ();
+  let pid = find_events_pid () in
+  create_and_free ~pid (); (* success *)
+  create_and_free (); (* success *)
+  make_unreadable ();
+  create_and_free ~pid (); (* fail, cannot open *)
+  create_and_free (); (* fail, cannot open *)
+  create_and_free (); (* fail, cannot open *)

--- a/testsuite/tests/lib-runtime-events/test_create_cursor_failures.reference
+++ b/testsuite/tests/lib-runtime-events/test_create_cursor_failures.reference
@@ -1,0 +1,6 @@
+Runtime_events: no ring for current process.          Was runtime_events started?
+OK
+OK
+Runtime_events: could not create cursor for specified path.
+Runtime_events: could not create cursor for specified path.
+Runtime_events: could not create cursor for specified path.

